packages/sdk-components-react/src/html-embed.tsx

```diff
@@ -158,26 +160,57 @@ type HtmlEmbedProps = {
 
 export const HtmlEmbed = forwardRef<HTMLDivElement, HtmlEmbedProps>(
   (props, ref) => {
-    const { renderer } = useContext(ReactSdkContext);
     const { code, executeScriptOnCanvas, clientOnly, ...rest } = props;
+    const { renderer } = useContext(ReactSdkContext);
+
+    const isServer = useSyncExternalStore(
+      () => () => {},
+      () => false,
+      () => true
+    );
+
+    const [ssrRendered] = useState(isServer);
 
     // - code can be actually undefined when prop is not provided
     // - cast code to string in case non-string value is computed from expression
     if (code === undefined || String(code).trim().length === 0) {
       return <Placeholder innerRef={ref} {...rest} />;
     }
 
-    const execute = shouldExecute({
-      renderer,
-      executeScriptOnCanvas,
-      clientOnly,
-    });
+    if (ssrRendered) {
+      // We are on published site, on server rendering or after hydration
+
+      if (clientOnly === false) {
+        return <ServerEmbed innerRef={ref} code={code} {...rest} />;
+      }
 
-    if (execute) {
-      return <ExecutableHtml innerRef={ref} code={code} {...rest} />;
+      return (
+        <ClientOnly>
+          <ClientEmbed innerRef={ref} code={code} {...rest} />
+        </ClientOnly>
+      );
+    }
+    // We are or on canvas | preview | published site after client routing
+
+    // The only case we need to prevent script execution if it's explicitly disabled on the canvas
+    if (renderer === "canvas" && executeScriptOnCanvas !== true) {
+      return (
+        <ClientOnly>
+          <ClientEmbedWithNonExecutableScripts
+            innerRef={ref}
+            code={code}
+            {...rest}
+          />
+        </ClientOnly>
+      );
     }
 
-    return <InnerHtml innerRef={ref} code={code} {...rest} />;
+    // Use key={code} to allow scripts to be reexecuted when code has changed
+    return (
+      <ClientOnly>
+        <ClientEmbed key={code} innerRef={ref} code={code} {...rest} />
+      </ClientOnly>
+    );
```

Comment:
please move this comment above key
```suggestion
    return (
      <ClientOnly>
        <ClientEmbed 
          // Use key={code} to allow scripts to be reexecuted when code has changed
          key={code} 
          innerRef={ref} code={code} {...rest} />
      </ClientOnly>
    );
```